src/matches.rs

```diff
@@ -209,3 +338,71 @@ fn match_template(cx: &LateContext,
         }
     }
 }
+
+fn overlapping<T>(ranges: &[SpannedRange<T>]) -> Option<(&SpannedRange<T>, &SpannedRange<T>)>
+    where T: Copy + Ord {
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+    enum Kind<'a, T: 'a> {
+        Start(T, &'a SpannedRange<T>),
+        End(T, &'a SpannedRange<T>),
+    }
+
+    impl<'a, T: Copy> Kind<'a, T> {
+        fn range(&self) -> &'a SpannedRange<T> {
+            match *self {
+                Kind::Start(_, r) | Kind::End(_, r) => r
+            }
+        }
+
+        fn value(self) -> T {
+            match self {
+                Kind::Start(t, _) | Kind::End(t, _) => t
+            }
+        }
+    }
+
+    impl<'a, T: Copy + Ord> PartialOrd for Kind<'a, T> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    impl<'a, T: Copy + Ord> Ord for Kind<'a, T> {
+        fn cmp(&self, other: &Self) -> Ordering {
+            self.value().cmp(&other.value())
+        }
+    }
+
+    let mut values = Vec::with_capacity(2*ranges.len());
+
+    for r in ranges {
+        values.push(Kind::Start(r.node.0, &r));
+        values.push(Kind::End(r.node.1, &r));
+    }
+
+    values.sort();
+
+    for (a, b) in values.iter().zip(values.iter().skip(1)) {
+        match (a, b) {
+            (&Kind::Start(_, ra), &Kind::End(_, rb)) => if ra.node != rb.node { return Some((ra, rb)) },
+            (&Kind::End(a, _), &Kind::Start(b, _)) if a != b => (),
+            _ => return Some((&a.range(), &b.range())),
+        }
+    }
+
+    None
+}
+
+#[test]
```

Comment:
We have a `tests/` folder in our project. Tests should go there (and not all of them need to be compile tests).